src/selectors/entities/channels.js

```diff
@@ -1156,3 +1156,44 @@ export const getSortedDirectChannelWithUnreadsIds: (GlobalState, Channel, boolea
         return filterChannels(unreadChannelIds, favoritePreferences, directChannelIds, false, favoritesAtTop);
     },
 );
+
+export const getDefaultChannelForTeams: (GlobalState) => RelationOneToOne<Team, Channel> = createSelector(
+    getAllChannels,
+    (channels: IDMappedObjects<Channel>): RelationOneToOne<Team, Channel> => {
+        const result = {};
+        for (const channel of Object.keys(channels).map((key) => channels[key])) {
+            if (channel && channel.name === General.DEFAULT_CHANNEL) {
+                result[channel.team_id] = channel;
+            }
+        }
+        return result;
+    }
+);
+
+export const getMyFirstChannelForTeams: (GlobalState) => RelationOneToOne<Team, Channel> = createSelector(
+    getAllChannels,
+    getMyChannelMemberships,
+    getCurrentUser,
+    (channels: IDMappedObjects<Channel>, members: RelationOneToOne<Channel, ChannelMembership>, currentUser: UserProfile): RelationOneToOne<Team, Channel> => {
+        const locale = currentUser.locale || General.DEFAULT_LOCALE;
+        const result: RelationOneToOne<Team, Channel> = {};
+        for (const channel of Object.keys(channels).map((key) => channels[key]).sort(sortChannelsByDisplayName.bind(null, locale))) {
+            if (!result[channel.team_id] && members[channel.id]) {
+                result[channel.team_id] = channel;
+            }
+        }
+        return result
+    }
+);
+
+export const getRedirectChannelNameForTeam = (state: GlobalState, teamId: string): string => {
+    const defaultChannels = getDefaultChannelForTeams(state);
+    const firstChannelForTeams = getMyFirstChannelForTeams(state);
+    const canIJoinPublicTeams = !hasNewPermissions(state) || haveITeamPermission(state, {team: teamId, permission: Permissions.JOIN_PUBLIC_CHANNELS});
+    const members = getMyChannelMemberships(state);
+
+    if ((!defaultChannels[teamId] || !members[defaultChannels[teamId].id]) && !canIJoinPublicTeams) {
```

Comment:
It's not immediately clear how to interpret this business logic. Could we simplify with early returns and comments?

Reply:
Sure